File: android/src/main/kotlin/com/example/native_cache_video_player/NativeCacheVideoPlayerPlugin.kt
package com.example.native_cache_video_player

import android.content.Context
import androidx.annotation.NonNull
import io.flutter.embedding.engine.plugins.FlutterPlugin
import io.flutter.plugin.common.EventChannel
import io.flutter.plugin.common.MethodCall
import io.flutter.plugin.common.MethodChannel
import io.flutter.plugin.common.MethodChannel.MethodCallHandler
import io.flutter.plugin.common.MethodChannel.Result
import io.flutter.view.TextureRegistry
import io.flutter.plugin.common.BinaryMessenger
import android.net.Uri
import java.util.concurrent.ConcurrentHashMap
import io.flutter.FlutterInjector
import android.os.Handler
import android.os.Looper

/**
 * The main entry point for the Native Cache Video Player Android implementation.
 *
 * This class implements the [FlutterPlugin] interface to register with the engine
 * and handles all MethodChannel communication from the Dart layer.
 */
class NativeCacheVideoPlayerPlugin: FlutterPlugin, MethodCallHandler {
    private lateinit var channel : MethodChannel
    private lateinit var textureRegistry: TextureRegistry
    private lateinit var context: Context
    private lateinit var binaryMessenger: BinaryMessenger

    private val players = LinkedHashMap<Long, FLTVideoPlayer>(16, 0.75f, true) // access-order LRU
    private var mixWithOthers = false
    private val MAX_PLAYERS = 4

    private val componentCallbacks = object : android.content.ComponentCallbacks2 {
        override fun onConfigurationChanged(newConfig: android.content.res.Configuration) {}
        override fun onLowMemory() {
            purgeInactivePlayers("onLowMemory")
        }
        override fun onTrimMemory(level: Int) {
            if (level >= android.content.ComponentCallbacks2.TRIM_MEMORY_RUNNING_CRITICAL ||
                level >= android.content.ComponentCallbacks2.TRIM_MEMORY_COMPLETE) {
                purgeInactivePlayers("onTrimMemory($level)")
            }
        }
    }

    private fun purgeInactivePlayers(reason: String) {
        println("NCVP: [MEMORY] $reason. Purging inactive players.")
        val toRemove = mutableListOf<Long>()
        synchronized(players) {
            for ((textureId, player) in players) {
                if (!player.isPlaying()) {
                    toRemove.add(textureId)
                }
            }
            for (textureId in toRemove) {
                println("NCVP: [MEMORY] Evicting idle player $textureId")
                players.remove(textureId)?.dispose()
            }
        }
    }

    override fun onAttachedToEngine(@NonNull flutterPluginBinding: FlutterPlugin.FlutterPluginBinding) {
        channel = MethodChannel(flutterPluginBinding.binaryMessenger, "native_cache_video_player")
        channel.setMethodCallHandler(this)
        textureRegistry = flutterPluginBinding.textureRegistry
        context = flutterPluginBinding.applicationContext
        binaryMessenger = flutterPluginBinding.binaryMessenger

        context.registerComponentCallbacks(componentCallbacks)

        // Initialize CacheManager on a background thread to avoid blocking UI
        Thread {
            CacheManager.getInstance(context)
        }.start()
    }

    override fun onMethodCall(@NonNull call: MethodCall, @NonNull result: Result) {
        when (call.method) {
            "init" -> {
                result.success(null)
            }
            "updateConfig" -> {
                val args = call.arguments as? Map<String, Any?>
                val max = (args?.get("maxCacheSize") as? Number)?.toLong() ?: (500 * 1024 * 1024L)
                // Concurrent parameters are ignored – kept only for API compatibility
                val concurrent = (args?.get("concurrentPrecache") as? Number)?.toInt() ?: 3
                val concurrentDuringPlaybackDownload = (args?.get("concPrecacheDuringPlaybackDownload") as? Number)?.toInt() ?: 3

                Thread {
                    CacheManager.getInstance(context).updateConfig(max, concurrent, concurrentDuringPlaybackDownload)
                    Handler(Looper.getMainLooper()).post {
                        result.success(null)
                    }
                }.start()
            }
            "create" -> {
                val args = call.arguments as? Map<String, Any?>
                val uri = args?.get("uri") as? String
                val formatHint = args?.get("formatHint") as? String
                val httpHeaders = args?.get("httpHeaders") as? Map<String, String> ?: emptyMap()

                if (uri == null) {
                    result.error("invalid_args", "uri is null", null)
                    return
                }

                val asset = args?.get("asset") as? String
                val packageName = args?.get("package") as? String

                val producer = textureRegistry.createSurfaceProducer()
                val textureId = producer.id()

                // Limit concurrent players
                synchronized(players) {
                    if (players.size >= MAX_PLAYERS) {
                        val eldestKey = players.keys.iterator().next()
                        println("NCVP: [CONCURRENCY] Evicting oldest player $eldestKey due to limit ($MAX_PLAYERS)")
                        players.remove(eldestKey)?.dispose()
                    }
                }

                val finalUri = if (asset != null) {
                    val loader = FlutterInjector.instance().flutterLoader()
                    val assetLookupKey = if (packageName != null) loader.getLookupKeyForAsset(asset, packageName) else loader.getLookupKeyForAsset(asset)
                    "asset:///$assetLookupKey"
                } else {
                    uri
                }

                val eventChannel = EventChannel(binaryMessenger, "native_cache_video_player/videoEvents$textureId")

                val player = FLTVideoPlayer(context, eventChannel, producer, finalUri, formatHint, httpHeaders)
                player.setAudioAttributes(mixWithOthers)
                synchronized(players) {
                    players[textureId] = player
                }

                val reply = HashMap<String, Any>()
                reply["textureId"] = textureId
                result.success(reply)
            }
            "setMixWithOthers" -> {
                mixWithOthers = call.argument<Boolean>("mixWithOthers") ?: false
                result.success(null)
            }
            "clearAllCache" -> {
                Thread {
                    CacheManager.getInstance(context).clearAllCache(result)
                }.start()
            }
            "removeFile" -> {
                val args = call.arguments as? Map<String, Any?>
                val url = args?.get("url") as? String
                if (url != null) {
                    Thread {
                        CacheManager.getInstance(context).removeFile(url, result)
                    }.start()
                } else {
                    result.error("invalid_args", "url is null", null)
                }
            }
            "enforceCacheLimit" -> {
                val args = call.arguments as? Map<String, Any?>
                val max = (args?.get("maxCacheSize") as? Number)?.toLong()
                if (max != null) {
                    Thread {
                        CacheManager.getInstance(context).enforceCacheLimit(max, result)
                    }.start()
                } else {
                    result.error("invalid_args", "maxCacheSize is missing", null)
                }
            }
            else -> {
                val args = call.arguments as? Map<String, Any?>
                val textureId = (args?.get("textureId") as? Number)?.toLong()

                if (textureId != null) {
                    if (call.method == "dispose") {
                        val pTextureId = textureId
                        val player = synchronized(players) { players.remove(pTextureId) }
                        player?.dispose()
                        result.success(null)
                        return
                    }

                    val player = synchronized(players) { players[textureId] } // LinkedHashMap.get updates access order if configured as such
                    if (player != null) {
                        when (call.method) {
                            "play" -> { player.play(); result.success(null) }
                            "pause" -> { player.pause(); result.success(null) }
                            "setLooping" -> {
                                val looping = args?.get("looping") as? Boolean ?: false
                                player.setLooping(looping)
                                result.success(null)
                            }
                            "setVolume" -> {
                                val volume = (args?.get("volume") as? Number)?.toDouble() ?: 0.0
                                player.setVolume(volume)
                                result.success(null)
                            }
                            "setPlaybackSpeed" -> {
                                val speed = (args?.get("speed") as? Number)?.toDouble() ?: 1.0
                                player.setPlaybackSpeed(speed)
                                result.success(null)
                            }
                            "setCaptionOffset" -> {
                                val offset = (args?.get("offset") as? Number)?.toLong() ?: 0L
                                player.setCaptionOffset(offset)
                                result.success(null)
                            }
                            "seekTo" -> {
                                val pos = (args?.get("position") as? Number)?.toInt() ?: 0
                                player.seekTo(pos)
                                result.success(null)
                            }
                            "position" -> {
                                result.success(player.getPosition())
                            }
                            else -> result.notImplemented()
                        }
                    } else {
                        result.error("unknown_player", "No player found for textureId $textureId", null)
                    }
                } else {
                    result.notImplemented()
                }
            }
        }
    }

    override fun onDetachedFromEngine(@NonNull binding: FlutterPlugin.FlutterPluginBinding) {
        channel.setMethodCallHandler(null)
        context.unregisterComponentCallbacks(componentCallbacks)
        synchronized(players) {
            for (player in players.values) {
                player.dispose()
            }
            players.clear()
        }
    }
}

File: ios/Classes/NativeCacheVideoPlayerPlugin.swift
import Flutter
import UIKit
import AVFoundation

/// The main plugin class for the Native Cache Video Player iOS implementation.
///
/// This class handles the registration of the plugin with the Flutter registrar 
/// and dispatches method calls from Dart to the appropriate player instances.
public class NativeCacheVideoPlayerPlugin: NSObject, FlutterPlugin {
    private let registrar: FlutterPluginRegistrar
    private var players = [Int64: FLTVideoPlayer]()
    private var playerOrder = [Int64]() // Track LRU order (end is most recent)
    private let playersQueue = DispatchQueue(label: "native_cache_video_player.playersQueue")
    private let MAX_PLAYERS = 4

    init(registrar: FlutterPluginRegistrar) {
        self.registrar = registrar
        super.init()
        setupMemoryWarningObserver()
    }

    public static func register(with registrar: FlutterPluginRegistrar) {
        let channel = FlutterMethodChannel(name: "native_cache_video_player", binaryMessenger: registrar.messenger())
        let instance = NativeCacheVideoPlayerPlugin(registrar: registrar)
        registrar.addMethodCallDelegate(instance, channel: channel)
    }

    private func setupMemoryWarningObserver() {
        NotificationCenter.default.addObserver(
            self,
            selector: #selector(handleMemoryWarning),
            name: UIApplication.didReceiveMemoryWarningNotification,
            object: nil
        )
    }

    @objc private func handleMemoryWarning() {
        playersQueue.async { [weak self] in
            guard let self = self else { return }
            print("NCVP: [MEMORY] Low memory warning. Purging all paused players.")
            
            // Dispose all players that are not currently playing
            // Note: We can only safely dispose on main thread if needed, 
            // but FLTVideoPlayer.dispose handles its own thread safety/cleanup.
            var keysToRemove = [Int64]()
            for (textureId, player) in self.players {
                if !player.isPlaying() {
                    keysToRemove.append(textureId)
                }
            }
            
            for textureId in keysToRemove {
                if let player = self.players.removeValue(forKey: textureId) {
                    self.playerOrder.removeAll { $0 == textureId }
                    print("NCVP: [MEMORY] Evicting idle player \(textureId)")
                    DispatchQueue.main.async {
                        self.registrar.textures().unregisterTexture(textureId)
                        player.dispose()
                    }
                }
            }
        }
    }

    /// Dispatches method calls from the Flutter layer to the appropriate handler.
    public func handle(_ call: FlutterMethodCall, result: @escaping FlutterResult) {
        switch call.method {
        case "init":
            result(nil)
        case "updateConfig":
            guard let args = call.arguments as? [String: Any] else {
                result(FlutterError(code: "invalid_args", message: "Arguments missing", details: nil))
                return
            }
            let max = args["maxCacheSize"] as? Int64 ?? (500 * 1024 * 1024)
            let concurrent = args["concurrentPrecache"] as? Int ?? 2
            let concurrentDuringPlaybackDownload = args["concPrecacheDuringPlaybackDownload"] as? Int ?? 0
            
            CacheManager.shared.updateConfig(maxSize: max, concurrent: concurrent, concurrentWhilePlaying: concurrentDuringPlaybackDownload)
            result(nil)
        case "create":
            guard let args = call.arguments as? [String: Any] else {
                result(FlutterError(code: "invalid_args", message: "Arguments are missing", details: nil))
                return
            }
            create(args: args, result: result)
        case "setMixWithOthers":
            guard let args = call.arguments as? [String: Any],
                  let mixWithOthers = args["mixWithOthers"] as? Bool else {
                result(FlutterError(code: "invalid_args", message: "mixWithOthers argument missing", details: nil))
                return
            }
            let session = AVAudioSession.sharedInstance()
            do {
                if mixWithOthers {
                    try session.setCategory(.playback, mode: .default, options: .mixWithOthers)
                } else {
                    try session.setCategory(.playback, mode: .default, options: [])
                }
                result(nil)
            } catch {
                result(FlutterError(code: "set_category_error", message: error.localizedDescription, details: nil))
            }
        case "clearAllCache":
            playersQueue.async {
                CacheManager.shared.clearAllCache { cleared, error in
                    DispatchQueue.main.async {
                        if let error = error {
                            result(FlutterError(code: "cache_error", message: "Failed to clear cache: \(error.localizedDescription)", details: nil))
                        } else {
                            result(cleared)
                        }
                    }
                }
            }
        case "enforceCacheLimit":
            if let args = call.arguments as? [String: Any], let max = args["maxCacheSize"] as? Int64 {
                playersQueue.async {
                    CacheManager.shared.enforceCacheLimit(maxSize: max) { removed, error in
                        DispatchQueue.main.async {
                            if let error = error {
                                result(FlutterError(code: "cache_error", message: "Failed to enforce cache limit: \(error.localizedDescription)", details: nil))
                            } else {
                                result(removed)
                            }
                        }
                    }
                }
            } else {
                 result(FlutterError(code: "invalid_args", message: "maxCacheSize missing", details: nil))
            }
        case "removeFile":
             if let args = call.arguments as? [String: Any], let url = args["url"] as? String {
                playersQueue.async {
                    CacheManager.shared.removeFile(for: url) { removed, error in
                        DispatchQueue.main.async {
                            if let error = error {
                                result(FlutterError(code: "cache_error", message: "Failed to remove file: \(error.localizedDescription)", details: nil))
                            } else {
                                result(removed)
                            }
                        }
                    }
                }
            } else {
                 result(FlutterError(code: "invalid_args", message: "url missing", details: nil))
            }
        default:
            guard let args = call.arguments as? [String: Any],
                  let textureId = args["textureId"] as? Int64 else {
                result(FlutterMethodNotImplemented)
                return
            }
            
            playersQueue.async { [weak self] in
                guard let self = self else { return }
                guard let player = self.players[textureId] else {
                    // Player was already disposed (e.g. by LRU eviction or memory pressure).
                    // Return success instead of an error – the caller's intent is satisfied.
                    print("NCVP: [WARN] No player found for textureId \(textureId) (already disposed). Ignoring \(call.method).")
                    DispatchQueue.main.async {
                        result(nil)
                    }
                    return
                }
                
                // Update LRU order
                self.playerOrder.removeAll { $0 == textureId }
                self.playerOrder.append(textureId)
                
                if call.method == "setCaptionOffset" {
                    if let offset = args["offset"] as? Int {
                        player.setCaptionOffset(offset)
                    }
                    DispatchQueue.main.async {
                        result(nil)
                    }
                    return
                }
                
                if call.method == "dispose" {
                    let removedPlayer = self.players.removeValue(forKey: textureId)
                    self.playerOrder.removeAll { $0 == textureId }
                    
                    if let player = removedPlayer {
                        DispatchQueue.main.async {
                            self.registrar.textures().unregisterTexture(textureId)
                            player.dispose() // now guaranteed on main thread
                        }
                    } else {
                        // Player already gone – just return success (no error)
                        DispatchQueue.main.async {
                            result(nil)
                        }
                    }
                    return
                }
                
                player.handle(call) { playerResult in
                    DispatchQueue.main.async {
                        result(playerResult)
                    }
                }
            }
        }
    }
    
    /// Creates a new player instance and registers it with the Flutter texture registry.
    private func create(args: [String: Any], result: @escaping FlutterResult) {
        guard let uri = args["uri"] as? String else {
             result(FlutterError(code: "invalid_args", message: "URI is required", details: nil))
             return
        }
        
        let assetName = args["asset"] as? String
        let packageName = args["package"] as? String
        let httpHeaders = args["httpHeaders"] as? [String: String] ?? [:]
        
        // Offload player setup to background
        playersQueue.async { [weak self] in
            guard let self = self else { return }
            
            // Limit concurrent players
            if self.players.count >= self.MAX_PLAYERS {
                if let eldestId = self.playerOrder.first {
                    print("NCVP: [CONCURRENCY] Evicting oldest player \(eldestId) due to limit (\(self.MAX_PLAYERS))")
                    if let eldestPlayer = self.players.removeValue(forKey: eldestId) {
                        self.playerOrder.removeFirst()
                        DispatchQueue.main.async {
                            self.registrar.textures().unregisterTexture(eldestId)
                            eldestPlayer.dispose()
                        }
                    }
                }
            }
            
            var finalUri = uri
            if let assetName = assetName {
                let assetKey: String
                if let packageName = packageName {
                    assetKey = self.registrar.lookupKey(forAsset: assetName, fromPackage: packageName)
                } else {
                    assetKey = self.registrar.lookupKey(forAsset: assetName)
                }
                
                if let path = Bundle.main.path(forResource: assetKey, ofType: nil) {
                    finalUri = URL(fileURLWithPath: path).absoluteString
                }
            }
            
            let player = FLTVideoPlayer(
                uri: finalUri,
                registrar: self.registrar,
                httpHeaders: httpHeaders
            )
            
            DispatchQueue.main.async {
                let textureId = self.registrar.textures().register(player)
                player.textureId = textureId
                player.setupEventChannel(textureId: textureId)
                
                self.playersQueue.async {
                    self.players[textureId] = player
                    self.playerOrder.append(textureId)
                    DispatchQueue.main.async {
                        result(["textureId": textureId])
                    }
                }
            }
        }
    }
}
